File: packages/zopiq_live_card/android/src/main/java/com/siteonlab/zopiq_live_card/ZopiqLiveCardPlugin.java
package com.siteonlab.zopiq_live_card;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.text.TextUtils;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.PluginRegistry;

/**
 * The bridge. Two jobs and nothing else: draw the card when Dart says the order moved, and tell Dart
 * which order a tap on the card was asking about.
 *
 * <p><b>Why this runs twice.</b> Two Flutter engines in this process register this plugin — the
 * app's own, and the one {@code firebase_messaging} spins up to run the background message handler
 * with the app killed. Both get their own instance and their own channel; only the app's is ever
 * attached to an activity. That is exactly the split the card needs: the background instance draws,
 * the foreground instance also listens for taps.
 */
public final class ZopiqLiveCardPlugin
        implements FlutterPlugin, MethodChannel.MethodCallHandler, ActivityAware,
                PluginRegistry.NewIntentListener {

    private static final String CHANNEL = "com.siteonlab.zopiq_live_card";

    private MethodChannel channel;
    private Context context;
    private Activity activity;

    /**
     * A tap that arrived before Dart was ready to hear about it.
     *
     * <p>A cold start from the lock screen runs the whole app between the tap and the first time
     * anything asks where it should navigate, so the order id waits here until {@code
     * consumeLaunchOrderId} collects it. Static because the activity's plugin instance is torn down
     * and rebuilt across a configuration change, and a rotation on the way in is not a reason to
     * lose the tap.
     */
    private static String pendingLaunchOrderId;

    // ---------------------------------------------------------------- FlutterPlugin

    @Override
    public void onAttachedToEngine(FlutterPluginBinding binding) {
        context = binding.getApplicationContext();
        channel = new MethodChannel(binding.getBinaryMessenger(), CHANNEL);
        channel.setMethodCallHandler(this);
    }

    @Override
    public void onDetachedFromEngine(FlutterPluginBinding binding) {
        if (channel != null) channel.setMethodCallHandler(null);
        channel = null;
        context = null;
    }

    // ---------------------------------------------------------------- calls from Dart

    @Override
    public void onMethodCall(MethodCall call, MethodChannel.Result result) {
        if (context == null) {
            result.error("no_context", "The plugin is not attached to an engine.", null);
            return;
        }

        switch (call.method) {
            case "show":
                LiveCardNotification.show(
                        context,
                        intArg(call, "id"),
                        stringArg(call, "orderId"),
                        stringArg(call, "title"),
                        stringArg(call, "body"),
                        stringArg(call, "subText"),
                        intArg(call, "progress"));
                result.success(null);
                break;

            case "cancel":
                LiveCardNotification.cancel(context, intArg(call, "id"));
                result.success(null);
                break;

            case "consumeLaunchOrderId":
                // Drain the activity's intent first: on a cold start the plugin can be attached to
                // the engine before it is attached to the activity, so the latch below may still be
                // empty when Dart asks.
                takeOrderIdFrom(activity == null ? null : activity.getIntent());
                final String pending = pendingLaunchOrderId;
                pendingLaunchOrderId = null;
                result.success(pending);
                break;

            default:
                result.notImplemented();
        }
    }

    // ---------------------------------------------------------------- taps

    @Override
    public boolean onNewIntent(Intent intent) {
        // The app was already running. There is a live channel, so hand it straight over rather
        // than parking it and waiting to be asked.
        final String orderId = readOrderId(intent);
        if (orderId != null && channel != null) {
            channel.invokeMethod("onOrderTapped", orderId);
        }
        // Never consume the intent: app_links and firebase_messaging read the same one.
        return false;
    }

    @Override
    public void onAttachedToActivity(ActivityPluginBinding binding) {
        activity = binding.getActivity();
        binding.addOnNewIntentListener(this);
        takeOrderIdFrom(activity.getIntent());
    }

    @Override
    public void onReattachedToActivityForConfigChanges(ActivityPluginBinding binding) {
        onAttachedToActivity(binding);
    }

    @Override
    public void onDetachedFromActivity() {
        activity = null;
    }

    @Override
    public void onDetachedFromActivityForConfigChanges() {
        onDetachedFromActivity();
    }

    private static void takeOrderIdFrom(Intent intent) {
        final String orderId = readOrderId(intent);
        if (orderId != null) pendingLaunchOrderId = orderId;
    }

    /**
     * Read the order id out of an intent and strike it off, so a resume or a rotation does not
     * re-navigate someone who has since walked elsewhere in the app.
     */
    private static String readOrderId(Intent intent) {
        if (intent == null) return null;
        final String orderId = intent.getStringExtra(LiveCardNotification.EXTRA_ORDER_ID);
        if (TextUtils.isEmpty(orderId)) return null;
        intent.removeExtra(LiveCardNotification.EXTRA_ORDER_ID);
        return orderId;
    }

    // ---------------------------------------------------------------- argument plumbing

    private static String stringArg(MethodCall call, String name) {
        return call.argument(name);
    }

    private static int intArg(MethodCall call, String name) {
        final Integer value = call.argument(name);
        return value == null ? 0 : value;
    }
}

File: packages/zopiq_live_card/ios/Classes/ZopiqLiveCardPlugin.swift
import Flutter
import UIKit

#if canImport(ActivityKit)
  import ActivityKit
#endif

/// The bridge, and the counterpart of `ZopiqLiveCardPlugin.java`. Same channel
/// name, same four methods, same arguments — so `zopiq_live_card.dart` needs no
/// branch of its own.
///
/// **What is genuinely different from Android.** Android draws a notification;
/// iOS runs a Live Activity, which is a separate process rendering a SwiftUI
/// view it owns. The consequences worth knowing:
///
///  * There is no integer notification id on iOS. Dart still passes one, because
///    Android needs it and one API is better than two, and this class keeps a
///    map from that id to the `Activity` it started. See `activities`.
///  * A Live Activity cannot be started while the app is suspended. Android's
///    card can be drawn from the FCM background isolate with the app killed;
///    iOS requires the process to be running and not suspended, which a
///    `content-available` push gives us for a few seconds. That is why
///    `send-notification` now sends an APNs block — without it iOS never wakes
///    at all and none of this runs. The stronger guarantee is a push-to-start
///    token, and it is deliberately not here yet; see IOS_HANDOVER.md.
///  * Nothing redraws the bar. Android runs a foreground service ticking every
///    twenty seconds; the SwiftUI view animates between two dates on its own.
public class ZopiqLiveCardPlugin: NSObject, FlutterPlugin {
  private static let channelName = "com.siteonlab.zopiq_live_card"

  /// The live activities this process started, keyed by the id Dart used.
  ///
  /// Static for the same reason `pendingLaunchOrderId` is static on Android: a
  /// second Flutter engine (the one `firebase_messaging` spins up for a
  /// background message) registers a second instance of this plugin, and both
  /// must see the same activities or a background tick would start a *new* card
  /// beside the one the app already has rather than updating it.
  private static var activities: [Int: Any] = [:]

  /// A tap that arrived before Dart was ready to hear about it. Drained once by
  /// `consumeLaunchOrderId`, exactly as on Android.
  private static var pendingLaunchOrderId: String?

  private var channel: FlutterMethodChannel?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(
      name: channelName, binaryMessenger: registrar.messenger())
    let instance = ZopiqLiveCardPlugin()
    instance.channel = channel
    registrar.addMethodCallDelegate(instance, on: channel)
    registrar.addApplicationDelegate(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "show":
      show(call.arguments as? [String: Any] ?? [:], result)
    case "cancel":
      cancel(call.arguments as? [String: Any] ?? [:], result)
    case "consumeLaunchOrderId":
      let pending = ZopiqLiveCardPlugin.pendingLaunchOrderId
      ZopiqLiveCardPlugin.pendingLaunchOrderId = nil
      result(pending)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // ------------------------------------------------------------------ show

  private func show(_ args: [String: Any], _ result: @escaping FlutterResult) {
    #if canImport(ActivityKit)
      guard #available(iOS 16.1, *) else {
        // Too old for a Live Activity. Not an error: the ordinary push
        // notification still arrives, which is precisely the fallback an
        // Android 13 phone gets when it cannot draw a promoted card.
        result(nil)
        return
      }
      guard ActivityAuthorizationInfo().areActivitiesEnabled else {
        // The user turned Live Activities off for this app in Settings. Their
        // call, and nothing to report.
        result(nil)
        return
      }

      let id = intArg(args, "id")
      let orderId = args["orderId"] as? String ?? ""
      guard !orderId.isEmpty else {
        result(nil)
        return
      }

      // Milliseconds since the epoch on the wire, as on Android — the standard
      // codec sizes a Dart int to whatever fits, so read a generic number
      // rather than casting to Int64 and waiting for the first small value.
      let state = ZopiqLiveCardAttributes.ContentState(
        title: args["title"] as? String ?? "Your order",
        body: args["body"] as? String,
        phase: args["phase"] as? String ?? "prep",
        windowStart: date(args, "windowStart"),
        windowEnd: date(args, "windowEnd"),
        // -1 is "no override" on the wire, because null would have to be
        // unboxed on the other side anyway. Same convention as Android.
        progress: { let p = intArg(args, "progress", -1); return p < 0 ? nil : p }()
      )

      // Reusing an id redraws in place, which is the whole contract of `show`.
      if let existing = ZopiqLiveCardPlugin.activities[id]
        as? Activity<ZopiqLiveCardAttributes>
      {
        Task {
          await existing.update(using: state)
          result(nil)
        }
        return
      }

      do {
        let activity = try Activity.request(
          attributes: ZopiqLiveCardAttributes(orderId: orderId),
          contentState: state
        )
        ZopiqLiveCardPlugin.activities[id] = activity
        result(nil)
      } catch {
        // The commonest cause is the app being suspended: ActivityKit refuses
        // to start an activity for a process that is not properly awake. A card
        // that failed to appear must never take the background isolate down
        // with it — the order is still fine, the lock screen is just emptier.
        NSLog("Live card not started: \(error.localizedDescription)")
        result(nil)
      }
    #else
      result(nil)
    #endif
  }

  // ---------------------------------------------------------------- cancel

  private func cancel(_ args: [String: Any], _ result: @escaping FlutterResult) {
    #if canImport(ActivityKit)
      guard #available(iOS 16.1, *) else {
        result(nil)
        return
      }
      let id = intArg(args, "id")
      guard
        let activity = ZopiqLiveCardPlugin.activities.removeValue(forKey: id)
          as? Activity<ZopiqLiveCardAttributes>
      else {
        // An id that never had a card. Safe, and says so — the Dart side calls
        // this speculatively on both phases every time an order ends.
        result(nil)
        return
      }
      Task {
        // `.immediate` because the order is over. The default leaves the card
        // on the lock screen for up to four hours, which for a delivered order
        // is not a grace period, it is litter.
        await activity.end(dismissalPolicy: .immediate)
        result(nil)
      }
    #else
      result(nil)
    #endif
  }

  // ------------------------------------------------------------------ taps

  /// A tap on the Live Activity opens the app on this URL. The widget builds it
  /// as `zopiq://order/<id>`; everything else on that scheme is ignored here so
  /// the app's other deep links keep working.
  public func application(
    _ application: UIApplication,
    open url: URL,
    options: [UIApplication.OpenURLOptionsKey: Any] = [:]
  ) -> Bool {
    guard url.scheme == "zopiq", url.host == "order" else { return false }
    let orderId = url.lastPathComponent
    guard !orderId.isEmpty, orderId != "/" else { return false }

    if let channel = channel {
      // The app is running and there is a live channel — hand it straight over
      // rather than parking it and waiting to be asked.
      channel.invokeMethod("onOrderTapped", arguments: orderId)
    } else {
      ZopiqLiveCardPlugin.pendingLaunchOrderId = orderId
    }
    // Never claim the URL: other plugins read the same one.
    return false
  }

  // ---------------------------------------------------- argument plumbing

  private func intArg(_ args: [String: Any], _ name: String, _ fallback: Int = 0) -> Int {
    (args[name] as? NSNumber)?.intValue ?? fallback
  }

  private func date(_ args: [String: Any], _ name: String) -> Date {
    let ms = (args[name] as? NSNumber)?.doubleValue ?? 0
    return Date(timeIntervalSince1970: ms / 1000)
  }
}
